notes/smart-pointers-1.md

```diff
@@ -0,0 +1,247 @@
+# Smart Pointers, Part 1: unique pointers
+
+Contributors: Edward Zhang
+
+---
+## The problem with raw pointers
+
+Raw pointers are tricky to work with due to a variety of reasons
+
+- Its declaration doesn't tell you if the pointer points to a single object or an array
+  - ... do you use `delete` or `delete[]`?
+- Its declaration doesn't tell you if the pointer owns the object it points to
+  - ... should you destroy the pointed-to object after you're done using the pointer?
+- Easy to cause dangling pointers (i.e., objects are destroyed but we still have pointers to them)
+- Easy to cause double frees
+- Easy to cause memory leaks
+
+---
+## Smart pointers to the rescue!
+
+Smart pointers can help us!
+
+A **smart pointer** is an object that stores a pointer to a heap-allocated object
+- Think of them as wrappers around raw pointers
+- Smart pointers look and behave like regular pointers
+
+However, smart pointers can automatically delete the object to which it points, at the right time
+
+Let's take a look at 3 kinds of smart pointers
+
```

Comment:
Lots of talk about memory management. We may want to consider restructuring things.

Reply:
yeah I'd argue memory management is the whole point of smart pointers, pun intended